Search

jjeffryes opened the pull request:
This PR adds search and discovery to the client.

Tests will be added in a patch or a separate PR in order to not hold up the review.

@morebrownies I added a default image (the same one used in the address bar) for the on-page search if the logo the search provider gives fails to load, which is happening currently with the default search provider.

I also added error modals if the search provider call fails. If the call is to a different provider than the current default (ie: you pasted in a http address), the error message will give you the option to try the same search term with your default provider.

If the error occurs with the user's default provider, they are given the option to open the settings modal and change it. On doing so, the search view will reload and try to use the corrected url.

You can test this by putting a bogus url into the address bar, and by changing the search provider in settings/general to something incorrect.

Closes #203

morebrownies commented:
Should the search page be functional @jjeffryes? I'm not seeing any listings.

<img width="1197" alt="screen shot 2017-04-06 at 8 16 23 pm" src="https://cloud.githubusercontent.com/assets/13066886/24784222/ee9cdb58-1b05-11e7-8317-a30add8ceda4.png">

jjeffryes commented:
@morebrownies it looks like you don't have an up to date copy of the PR? That, or something is breaking during the page load. Do you see any errors in the console?

I just tried it on a new machine, and it is working normally. You should see this when clicking discover:

![image](https://cloud.githubusercontent.com/assets/1584275/24802065/88f935e8-1b74-11e7-9e01-3d5a7cdb646a.png)

rmisio commented:
When there's no results, the previous arrow shouldn't be active:

![image](https://cloud.githubusercontent.com/assets/794517/24807457/19e37d32-1b6d-11e7-8dae-c4de03ef0a31.png)

jjeffryes commented:
The issue Mike saw was he had run this branch at least once before when the url for the default was different, and that was saved to his settings.

I've updated the code to account for that, I'll push the fix after reviews are in.

jjeffryes commented:
I've added a state for no results found, and fixed the issue with the previous page button being active if no results are found. I'll push that code after reviews are completed so I don't disrupt anything in progress.

![image](https://cloud.githubusercontent.com/assets/1584275/24807641/f88bf992-1b86-11e7-94ba-eedfcd2b3d5e.png)

morebrownies commented:
Ok thanks @jjeffryes I'm now seeing results. 

None of the listings seem to load when I click them though and the filtering/sorting also doesn't seem to work.

morebrownies commented:
Also, we probably shouldn't display the suggestions label if there aren't any suggestions

![screen shot 2017-04-07 at 9 16 58 am](https://cloud.githubusercontent.com/assets/13066886/24809006/fbed705c-1b72-11e7-9e1e-250db2974545.png)

morebrownies commented:
In settings, I have my currency set to USD, yet the search results display in BTC.

![screen shot 2017-04-07 at 9 18 36 am](https://cloud.githubusercontent.com/assets/13066886/24809096/49968ad2-1b73-11e7-8e77-ba1e8182e9d6.png)

![screen shot 2017-04-07 at 9 19 37 am](https://cloud.githubusercontent.com/assets/13066886/24809129/5eb0957a-1b73-11e7-8d15-1fc26559b03d.png)

jjeffryes commented:
Re: suggestions, those are just stubbed in, I don't think they exist yet in the search server code.

I can comment them out for now.

morebrownies commented:
Also @jjeffryes is it possible to display the user's avatar? Right now it appears you're double outputting the listing image. 

![screen shot 2017-04-07 at 9 31 40 am](https://cloud.githubusercontent.com/assets/13066886/24809855/0bc6193c-1b75-11e7-814d-820ad3029007.png)

jjeffryes commented:
@morebrownies good catch on the currency, there was an issue with the server returning 'btc' instead of 'BTC'. I have that fixed, will push it when Rob's review is done.

The avatar being a listing image is a server issue, Tyler is working on it.

jjeffryes commented:
Tyler's working on suggestions, when he has them available I'll code them in. That might be a later PR.

morebrownies commented:
Ok cool I found a listing that loads @jjeffryes. I noticed when I click the listing, the overlay does not display the Store name / view store link. We should definitely get that outputting otherwise it's not clear what store you're looking at.

**Current**
![screen shot 2017-04-07 at 9 51 03 am](https://cloud.githubusercontent.com/assets/13066886/24810675/00051d8e-1b78-11e7-8f0f-1c5a850b118f.png)

**Expected**
![screen shot 2017-04-07 at 9 52 47 am](https://cloud.githubusercontent.com/assets/13066886/24810676/001a1c84-1b78-11e7-9d0a-11b030b580c5.png)

morebrownies commented on js/collections/Results.js:
```diff
@@ -0,0 +1,54 @@
+import { integerToDecimal } from '../utils/currency';
```
> Seems like some spacing is needed between the icon and text:

Glad you bought it up. I feel the same way. It looks like the spaces are in the DOM, but they're not outputting for some reason. 

> Should we be allowing search within this drop down?

Yeah, not super useful. In most cases, the dropdown would probably be under 5 options.

> The left border of the search button is not changing on hover:

If it's easier to put a gap between the input and button, that should be fine too.

jjeffryes commented:
The spacing on the icons in the filters looks correct on both my Mac and my Windows machine, can you think of a reason it would be different on yours, Rob?

<img width="306" alt="screenshot 2017-04-07 15 39 17" src="https://cloud.githubusercontent.com/assets/1584275/24816860/73cfe7f0-1ba8-11e7-841c-b9dacdc1a5d4.png">

jjeffryes commented:
In Windows the flag for Germany renders as just "DE" and the USA as "US". I'm not sure how much we can do about that, it shows up that way in the json when I look at the same text in a browser. The spacing is the same issue, the icons and text are whatever the search provider sends us. 

![image](https://cloud.githubusercontent.com/assets/1584275/24817069/13f60804-1ba9-11e7-9028-b9b3b3126c8f.png)

jjeffryes commented:
Any filters shown are what the server returns, right now I believe it's returning a default set with a blank search request. The client shouldn't intervene in any way in the filters, whatever the server sends should be assumed to be correct, since there's no way to know what filters a search provider will want to send.

jjeffryes commented:
All of the issues noticed in the review should be corrected now. Let me know if I missed anything.

morebrownies commented:
Ok testing now

morebrownies commented:
Store name not displaying @jjeffryes 

![screen shot 2017-04-07 at 2 33 51 pm](https://cloud.githubusercontent.com/assets/13066886/24820714/49b32b16-1b9f-11e7-9cea-4f545c5f0a65.png)

jjeffryes commented:
@morebrownies that's correct. The search provider doesn't send the store name currently, until it does it won't display (I mentioned that in one of the 300 comments above. ;) )

morebrownies commented:
ah, ok definitely missed that. my bad

morebrownies commented:
Hey @jjeffryes sorry found one more small detail. On a user's store, when no search results are found, we display the following:

![screen shot 2017-04-07 at 3 31 49 pm](https://cloud.githubusercontent.com/assets/13066886/24822110/958596e8-1ba7-11e7-9e9a-e8aeade15c32.png)

In search/discover, we display:
![screen shot 2017-04-07 at 3 33 02 pm](https://cloud.githubusercontent.com/assets/13066886/24822111/959af0e2-1ba7-11e7-999d-105c49449421.png)

It would be nice if those could match up.

tyler-smith commented:
As noted above, there are a few places the search server isn't working as expected, however I've reviewed the related code in the client and everything appears okay, so when I get the backend issues shored up it should be all good to go.

rmisio commented on js/views/search/Search.js:
```diff
@@ -0,0 +1,248 @@
+          dataType: 'json',
+        })
+        .done((data, status, xhr) => {
+        // make sure minimal data is present
```
Could we also only do the done() handler logic if the call isn't aborted?

rmisio replied:
Can we also abort the fetch in this View's remove?

jjeffryes replied:
Is it possible for done to fire and have a statusText of abort? If the call was aborted, wouldn't it fail instead?

jjeffryes replied:
(it looks like it's possible in the W3C spec, though it seems like it would be kind of a rare thing to happen)

rmisio replied:
Yeah, i didn't think that through. We could probably do without the abort check in the done handler.

rmisio commented on js/views/search/Search.js:
```diff
@@ -0,0 +1,248 @@
+  processTerm(term) {
+    this.term = term || '';
+    // if term has spaces, replace them with +
+    // if term is false, search for *
```
I think you want to be running these variables through encodeURIComponent() - spaces are just one thing to be handled as there are other non-url friendly characters that need to be encoded. To be safe I would apply this to all these vars, not just term, e.g:

```
  const query = `q=${term ||  '*'}`;
  const page = `&p=${this.serverPage}&ps=${this.pageSize}`;
  const encodedQueryString = encodeURIComponent(`${query}${this.sortByQuery}${this.filterQuery}${page}`);
  const searchURL = `${this.sProvider}?${encodedQueryString}`;
```

rmisio commented on js/views/search/Results.js:
```diff
@@ -0,0 +1,167 @@
+      const newPageCol = new ResultsCol();
+      this.pageCollections[page] = newPageCol;
+
+      newPageCol.fetch({
```
Can we give the same "abort" treatment here as in the callSearchProvider call. Basically, we only want the most recent loadPage to be happening, so:

```
if (this.newPageFetch) this.newPageFetch.abort();

this.newPageFetch = newPageCol.fetch()...
```

Then we'd want to "opt-out" of the handlers if the call has been aborted.

Finally, in this View's remove can we please abort any this.newPageFetch calls?

rmisio commented on js/utils/templateHelpers.js:
```diff
@@ -4,11 +4,17 @@ import {
   return app.polyglot.t(...args);
 }
 
+export function parseEmojis(msg) {
```
Minor, but can we change the argument to 'text'. 'msg' was appropriate in the chat view because it was parsing a chat message, but since we've added this to be used generically, I think the argument name should follow suit.

rmisio commented on js/views/search/Search.js:
```diff
@@ -0,0 +1,248 @@
+    }).render().open();
+    this.listenTo(errorDialog, 'click-changeProvider', () => {
+      this.changeProvider();
+    errorDialog.close();
```
Looks like some indentation is off here and also the `errorDialog.close` a few lines down.

rmisio commented on styles/components/_listingsGrid.scss:
```diff
@@ -1,15 +1,65 @@
 .listingsGrid {
```
Do you think it's worth it to put a comment in this file that these styles are applied to both the Search and Store pages, so please be mindful of both when making any changes? Your call.

jjeffryes commented on js/views/search/Search.js:
```diff
@@ -0,0 +1,229 @@
+    this.usingDefault = this.sProvider === app.localSettings.get('searchProvider');
+
+    // if not using a passed in URL, update the default provider if it changes
+    this.listenTo(app.localSettings, 'change:searchProvider', (_, provider) => {
```
Sure. This was just a copy paste from something else that used an underscore there.
